Added the 2 basic eBPF programs

trishms opened the pull request:
Includes new hook entry function prototypes (flow_md_t and mac_md_t)
@lianghaoGao

trishms committed:
Added the 2 basic eBPF programs

Includes new hook entry function prototypes (flow_md_t and mac_md_t)

lianghaoGao commented on tests/sample/data_usage/associatetoflow.c:
```diff
@@ -0,0 +1,27 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include "ebpf_helpers.h"
+#include "ebpf_nethooks.h"
```
I believe you need to combine two ebpf programs together so that they can share the maps.

trishms replied:
We only needed to combine the programs together so that the second eBPF program can access the flowId from the first eBPF program. Since we are doing a 2-map approach, combining is not necessary anymore.

lianghaoGao commented on tests/sample/data_usage/associatetoflow.c:
```diff
@@ -0,0 +1,27 @@
+    .max_entries = 500};
+
+
+#pragma clang section text = "flow"
```
You need same prefix to share maps

trishms replied:
Same prefix is needed if both maps are in same ELF file, in this case, we are not combining the programs because we are doing a 2-map approach. The eBPF programs do not need to access each other's maps.

lianghaoGao replied:
Got it. U r right

lianghaoGao commented on tests/sample/data_usage/countbytes.c:
```diff
@@ -0,0 +1,30 @@
+#pragma clang section data = "maps"
+ebpf_map_definition_t byte_map = {
+     .size = sizeof(ebpf_map_definition_t),
+    .type = EBPF_MAP_TYPE_HASH,
```
Isn't array map faster that hash map?

lianghaoGao commented on tests/sample/data_usage/countbytes.c:
```diff
@@ -0,0 +1,30 @@
+CountBytes(mac_md_t* ctx)
+{
+
+    uint64_t byte_count = ebpf_map_lookup_element(&byte_map, &ctx->five_tuple);
```
This returns pointer

lianghaoGao commented on tests/sample/data_usage/countbytes.c:
```diff
@@ -0,0 +1,30 @@
+{
+
+    uint64_t byte_count = ebpf_map_lookup_element(&byte_map, &ctx->five_tuple);
+    if (byte_count == NULL) {
```
just if (!byte_count)

lianghaoGao commented on tests/sample/data_usage/countbytes.c:
```diff
@@ -0,0 +1,30 @@
+
+#pragma clang section data = "maps"
+ebpf_map_definition_t byte_map = {
+     .size = sizeof(ebpf_map_definition_t),
```
extra space

lianghaoGao commented on tests/sample/data_usage/associatetoflow.c:
```diff
@@ -0,0 +1,27 @@
+    else { //flow deletion
+        ebpf_map_delete_element(&app_map, &ctx->five_tuple);
+    }
+Done:
```
You don't need the "Done" path, right?

lianghaoGao commented on tests/sample/data_usage/associatetoflow.c:
```diff
@@ -0,0 +1,27 @@
+AssociateFlowToContext(flow_md_t* ctx)
+{
+    if (ctx->flow_established_flag) {
+        ebpf_map_update_element(&app_map, &ctx->five_tuple, &ctx->app_id, 0);
```
Does the existing framework support testing map operations without having the program info provider?

lianghaoGao commented:
Can you move the user mode app under tools\ folder?

trishms committed:
Incorporated Eric's feedback for eBPF programs

lianghaoGao commented on include/ebpf_nethooks.h:
```diff
@@ -74,3 +75,29 @@ typedef enum _bind_action
+
+//Flow Established Hook
+typedef struct _flow_md
+{
```
Stay consistent on the format

lianghaoGao commented on include/ebpf_nethooks.h:
```diff
@@ -74,3 +75,29 @@ typedef enum _bind_action
+    uint16_t source_port; 
+    uint16_t dest_port;
+    uint8_t protocol;
+} five_tuple_t ;
```
extra space

lianghaoGao commented on tests/sample/data_usage/countbytes.c:
```diff
@@ -0,0 +1,26 @@
+#pragma clang section text = "mac"
+int
+CountBytes(mac_md_t* ctx)
+{
```
Does this compile at all? You're missing the return

trishms replied:
Woops, I took out Done and forgot to add in the return. Thanks Eric

trishms committed:
Incorporated Eric's Feedback Pt.2

trishms committed:
Reformatted to match epf-for-data-usage repo

Current on stand-by waiting to see whether should continue using forked repo or new repo with submodule since the extension is not fully completed and code cannot be separated entirely from ebpf-for-windows. Thus, moving recent changed in ebpf-for-data-usage here.
